packages/panna-sdk/src/core/utils/activity.ts

```diff
@@ -0,0 +1,454 @@
+ * //         owner: null,
+ * //         tokenInfo: {
+ * //           address: '0x...',
+ * //           name: 'Lisk of Life',
+ * //           symbol: 'LOL',
+ * //           decimals: null,
+ * //           type: 'erc-721',
+ * //           icon: null
+ * //         }
+ * //       },
+ * //     status: 'ok'
+ * //   }, {
+ * //     activityType: 'minted',
+ * //     transactionID: '0x...',
+ * //     amount: {
+ * //       type: 'erc-1155',
+ * //       tokenId: '29900446396079726096...',
+ * //       value: '100000000000',
+ * //       tokenInfo: {
+ * //         address: '0x...',
+ * //         name: 'Rarible',
+ * //         symbol: RARI,
+ * //         decimals: null,
+ * //         type: 'erc-1155',
+ * //         icon: null
+ * //       }
+ * //     },
+ * //     status: 'error'
+ * //   }, ...],
+ * //   metadata: { count: 10, offset: 0, limit: 10 }
+ * // }
+ * ```
+ */
+export const getActivity = async function (
+  params: GetActivityParams
+): Promise<GetActivityResult> {
+  if (!isValidAddress(params.address)) {
+    throw new Error('Invalid address format');
+  }
+
+  // Set default pagination params
+  const {
+    address,
+    offset = DEFAULT_PAGINATION_OFFSET,
+    limit = DEFAULT_PAGINATION_LIMIT
+  } = params;
+  const baseTxRequestUrl = `https://blockscout.lisk.com/api/v2/addresses/${address}/transactions`;
+
+  const cacheKeyTransactions = address;
+  const cacheKeyNextPageParams = `${address}_params`;
+
+  const userTransactions: BlockscoutTransaction[] = activityCache.has(
+    cacheKeyTransactions
+  )
+    ? (activityCache.get(cacheKeyTransactions) as BlockscoutTransaction[])
+    : [];
+
+  let nextPageParams: BlockscoutNextPageParams | null = activityCache.has(
+    cacheKeyNextPageParams
+  )
+    ? (activityCache.get(cacheKeyNextPageParams) as BlockscoutNextPageParams)
+    : null;
+
+  do {
+    const requestUrl =
+      nextPageParams === null
+        ? baseTxRequestUrl
+        : `${baseTxRequestUrl}?block_number=${nextPageParams.block_number}&index=${nextPageParams.index}&items_count=${nextPageParams.items_count}`;
+
+    const response: BlockscoutTransactionsResponse =
+      await httpUtils.request(requestUrl);
+
+    if ('code' in response && 'message' in response) {
+      throw new Error(`Cannot fetch user activities: ${response.message}`);
+    }
+
+    userTransactions.push(
+      ...(await fillTokenTransactions(address, response.items))
+    );
+    nextPageParams = response.next_page_params;
+
+    activityCache.set(cacheKeyTransactions, userTransactions);
+    activityCache.set(cacheKeyNextPageParams, nextPageParams);
+
+    // Continue the loop only if more pages exist
+    // Second condition ensures handling of no existing activities
+    if (
+      nextPageParams === null &&
+      userTransactions.length >= response.items.length
+    ) {
+      activityCache.delete(cacheKeyNextPageParams);
+      break;
+    }
+  } while (userTransactions.length <= offset + limit);
```

Comment:
if no more pages are left then we should also stop the loop?
```suggestion
  } while (userTransactions.length <= offset + limit || nextPageParams !== null);
```

Reply:
The if block above takes care of this.

This condition is unnecessary and will result in making unnecessary additional API requests until all the pages are requested and could result in hitting the API rate-limits.